src/toxcore/dht/server/mod.rs

```diff
@@ -208,9 +211,17 @@ impl Server {
             lan_discovery_enabled: true,
             is_ipv6_enabled: false,
             initial_bootstrap: Vec::new(),
+            precomputed_keys: Arc::new(RwLock::new(HashMap::new())),
         }
     }
 
+    /// Get PrecomputedKey of a peer or newly make and save it.
+    fn get_precomputed(&self, pk: PublicKey) -> PrecomputedKey {
+        let mut precomputed = self.precomputed_keys.write();
```

Comment:
Well, better take `read` lock first, then `write` to insert missing precomputed key.

Reply:
The `entry()` is good for HashMap on this purpose.
One read lock and one write lock can not be used in same scope.

Reply:
Then you should end the scope or `read` lock to get `write` lock.

Reply:
done